### ISSUE TITLE
Template incorrectly infers array/list

### ISSUE THREAD

**Issue by kkmuffme** [vimeo/psalm#8786]
https://psalm.dev/r/80a94f46a3

Afaik both errors are wrong and shouldn't be reported?

**Comment by boesing** [comment]
`string[]` is interpreted as `array<array-key,string>` and therefore its not a list.

Thats why there is no error reported when you pass a map in your 2nd example. Its only arguing about the return statement as it is not really clear.


In your last example, you end up having:

```php
$array = array( 'foo' => 'a', 'bar' => 'b', 0 => 'hello' );
```

Sometimes it helps when you actually use `var_dump`.

What are you actually trying to achieve with this method and why do you even need generic here?

**Comment by weirdan** [comment]
Given the `@param T` & `@return T` Psalm expects the function to return the same type. Psalm types are often very specific, e.g. when `list{'a','b'}` is passed, it's expected that `list{'a','b'}` is returned: https://psalm.dev/r/10ddbdaef6

**Comment by weirdan** [comment]
>  However if it passes the of string[] check without error, then it shouldn't emit an error for the return type either, if the return type matches the of string[] type.

`of string[]` defines the upper bound of the type. The type itself could be narrowed further, as in `list{'a', 'b'}` example.

In other words, when `foo()` is called with `list{'a','b'}` it effectively becomes https://psalm.dev/r/55bef429b2 (note that the list type is sealed here). And `list{0: 'a', 1: 'b', 2?: 'hello'}` is not a subtype of (sealed) `list{0: 'a', 1: 'b'}`.
